assets/scripts/segments/Building.jsx

```diff
@@ -0,0 +1,188 @@
+  }
+
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      variant: (props.position === 'left') ? 'leftBuildingVariant' : 'rightBuildingVariant',
+      height: (props.position === 'left') ? 'leftBuildingHeight' : 'rightBuildingHeight',
+      oldBuildingEnter: true,
+      newBuildingEnter: false,
+      switchBuildings: false
+    }
+  }
+
+  componentDidUpdate (prevProps, prevState) {
+    const { street, position } = this.props
+    const { variant, height } = this.state
+
+    const lastOverflow = (prevProps.street.remainingWidth < 0)
+    const streetOverflow = (street.remainingWidth < 0)
+
+    if (prevProps.street[height] !== street[height] || lastOverflow !== streetOverflow) {
+      createBuilding(this.streetSectionBuilding, street[variant], position, street[height], street)
+    }
+
+    if (prevProps.street[variant] && prevProps.street[variant] !== street[variant]) {
+      this.handleBuildingSwitch()
+    }
+
+    if (prevState.switchBuildings !== this.state.switchBuildings) {
+      this.props.calculateBuildingPerspective(this.oldStreetSectionBuilding)
+      this.props.calculateBuildingPerspective(this.streetSectionBuilding)
+      this.oldStreetSectionBuilding.classList.remove('hover')
+      createBuilding(this.streetSectionBuilding, street[variant], position, street[height], street)
+    }
+  }
+
+  onBuildingMouseEnter = (event) => {
+    window.addEventListener('keydown', this.handleKeyDown)
+    let type
+    if (this.props.position === 'left') {
+      type = INFO_BUBBLE_TYPE_LEFT_BUILDING
+    } else if (this.props.position === 'right') {
+      type = INFO_BUBBLE_TYPE_RIGHT_BUILDING
+    }
+
+    infoBubble.considerShowing(event, this.streetSectionBuilding, type)
+    resumeFadeoutControls()
+  }
+
+  onBuildingMouseLeave = (event) => {
+    window.removeEventListener('keydown', this.handleKeyDown)
+    // if (event.pointerType !== 'mouse') return
+    infoBubble.dontConsiderShowing()
+  }
+
+  handleKeyDown = (event) => {
+    const negative = (event.keyCode === KEYS.MINUS) ||
+      (event.keyCode === KEYS.MINUS_ALT) ||
+      (event.keyCode === KEYS.MINUS_KEYPAD)
+
+    const positive = (event.keyCode === KEYS.EQUAL) ||
+      (event.keyCode === KEYS.EQUAL_ALT) ||
+      (event.keyCode === KEYS.PLUS_KEYPAD)
+
+    const variant = this.props.street[this.state.variant]
+    const hasFloors = BUILDINGS[variant].hasFloors
+
+    if (negative && hasFloors) {
+      this.props.removeBuildingFloor(this.props.position)
+    } else if (positive && hasFloors) {
+      this.props.addBuildingFloor(this.props.position)
+    }
+
+    event.preventDefault()
+  }
+
+  handleChangeInRefs = (ref, isOldBuilding) => {
+    if (!this.state.switchBuildings && !isOldBuilding) return
+
+    if (this.state.switchBuildings && isOldBuilding) {
+      this.oldStreetSectionBuilding = ref
+    } else {
+      this.streetSectionBuilding = ref
+    }
+  }
+
+  handleBuildingSwitch = () => {
```

Comment:
Similarly here, even though this is in response to an event listener (`onEntered`) it should either be named as `handleEntered` (using the name of the event in the function name) or it can just be `switchBuildings` (which is all it does, it's just passed directly to a event listener).